### PR TITLE
Nft version 2 

### PR DESCRIPTION
**Pull request by rdinicut**
some changes from slither
 - update solidity compiler
 - removed the mandatoryFields from UserMintableERC721 initialize methods and now they are set in the subclass

### CODE REVIEW

File: contracts/erc721/PaymentObligation.sol [hunk 1]
```diff
@@ -127,47 +126,73 @@ contract PaymentObligation is Initializable, UserMintableERC721 {
...
 
     // Check if document has a read rule defined
     bytes8 readRoleIndex = super._requireReadRole(
       merkleRoot_,
       properties[0],
-      values[3],
-      salts[5],
-      proofs[5]
+      values[4],
+      salts[7],
```
Comment (manuelpolzhofer) [inline contracts/erc721/PaymentObligation.sol]: I would like to have constants for everything. No indexing with numbers it unnecessary complicated to read in my opinion.

File: contracts/erc721/PaymentObligation.sol [hunk 2]
```diff
@@ -127,47 +126,73 @@ contract PaymentObligation is Initializable, UserMintableERC721 {
...
 
     // Check if the token has the read role assigned to it
     super._requireTokenHasRole(
       merkleRoot_,
       tokenId,
       properties[1],
-      values[3], // the value from read role proof
-      salts[7],
-      proofs[7]
+      values[4], // the value from read role proof
```
Comment (manuelpolzhofer) [inline contracts/erc721/PaymentObligation.sol]: something like
```suggestion
      values[READ_RULE_VALUE_IDX], // the value from read role proof
```

File: contracts/erc721/PaymentObligation.sol [hunk 3]
```diff
@@ -127,47 +126,73 @@ contract PaymentObligation is Initializable, UserMintableERC721 {
...
+      salts[9],
```
Comment (manuelpolzhofer) [inline contracts/erc721/PaymentObligation.sol]:
```suggestion
      salts[READ_RULE_MAIN_IDX],
```

File: contracts/erc721/PaymentObligation.sol [hunk 4]
```diff
@@ -127,47 +126,73 @@ contract PaymentObligation is Initializable, UserMintableERC721 {
...
+      proofs[9]
```
Comment (manuelpolzhofer) [inline contracts/erc721/PaymentObligation.sol]:
```suggestion
      proofs[READ_RULE_MAIN_IDX]
```

**Comment by manuelpolzhofer** [comment]
a proposal for the idx constants in the proofs
```
const TOKENPROOF_IDX_SALT = 1
const TOKENPROOF_IDX_VALUE = 1
const TOKENPROOF_IDX_PROOF = 4
```

**Comment by manuelpolzhofer** [comment]
For avoiding to many constants we could define an array constant for each merkle proof.
```javascript
uint[4] constant TOKEN_ROLE_CONSTANT = [2, 5, 6, 5] // [property_idx, value_idx, proof_idx, salt_idx]
const PROPERTY_IDX = 0
const VALUE_IDX = 1
const PROOFS_IDX = 2
const SALT_IDX = 3
```
a index would be -1 if not used.
it would replace the current implementation
```javascript
salts[4]
```
to
```javascript
salts[TOKEN_ROLE_CONSTANT[SALT_IDX]]
```
now it would explain to the reader what is going on but maybe it is a bit over engineering.
 @rdinicut @lucasvo @pstehlik   what do you think?

**Comment by rdinicut** [comment]
One solution might be
/** @dev indexes used for the mint method arrays
  * Properties, values, salts and proofs have variant lengths and
  * not all arrays contain values for all the fields
  * The properties array contains only READ_ROLE and TOKEN_ROLE
  * The values array contains only GROSS_AMOUNT, CURRENCY,DUE_DATE, SENDER, READ_ROLE
  * salts and proofs contains values for all fields
  */
  uint8 constant internal GROSS_AMOUNT_IDX = 0;
  uint8 constant internal CURRENCY_IDX = 1;
  uint8 constant internal DUE_DATE_IDX = 2;
  uint8 constant internal SENDER_IDX = 3;
  uint8 constant internal READ_ROLE_IDX = 4;
  uint8 constant internal READ_ROLE_ACTION_IDX = 5;
  uint8 constant internal TOKEN_ROLE_IDX = 6;
  uint8 constant internal STATUS_IDX = 7;
  uint8 constant internal NEXT_VERSION_IDX = 8;
  uint8 constant internal NFT_UNIQUE_IDX = 9;

  uint8 constant internal READ_ROLE_PROP_IDX = 0;
  uint8 constant internal TOKEN_ROLE_PROP_IDX = 1;

**Pull request merged**